inference/sources/fly/main.go

```diff
@@ -0,0 +1,180 @@
+		Response         string    `json:"response"`
+		Timestamp        time.Time `json:"timestamp"`
+		Metadata         struct {
+			Text string `json:"type"`
+		} `json:"Metadata"`
+		Pageid              string                     `json:"pageid"`
+		Translated_response string                     `json:"translated_response"`
+		Hidden              map[string]json.RawMessage `json:"hidden"`
+	} `json:"items"`
+}
+
+type flyConnector struct {
+	BaseUrl      string `env:"FLY_BASE_URL,required"`
+	Key          string `env:"FLY_KEY,required"`
+	PageSize     int    `env:"FLY_PAGE_SIZE,required"`
+	ClientId     string `env:"FLY_CLIENT_ID,required"`
+	ClientSecret string `env:"FLY_CLIENT_SECRET,required"`
+}
+
+type FlyError struct {
+	Code string `json:"code"`
+}
+
+func (e *FlyError) Empty() bool {
+	return e.Code == ""
+}
+
+func (e *FlyError) Error() string {
+	return e.Code
+}
+
+func (c flyConnector) loadEnv() flyConnector {
+	err := env.Parse(&c)
+	handle(err)
+	return c
+}
+
+func Call(client *http.Client, baseUrl string, key string, form string, params *GetResponsesParams) (*GetResponsesResponse, error) {
+	sli := sling.New().Client(client).Base(baseUrl).Set("Accept", "application/json").Set("Authorization", fmt.Sprintf("Bearer %s", key))
+	res := new(GetResponsesResponse)
+	apiError := new(FlyError)
+
+	// fmt.Println("key ->", key)
+	// fmt.Println("params ->", params)
+	_, err := sli.Get("all?").QueryStruct(params).Receive(res, apiError)
+	// fmt.Println("resp: ", resp)
+	if err != nil {
+		return nil, err
+	}
+
+	if !apiError.Empty() {
+		return nil, apiError
+	}
+	return res, nil
+}
+
+func (c flyConnector) GetResponses(source *Source, form string, token string, idx int) <-chan *InferenceDataEvent {
+	events := make(chan *InferenceDataEvent)
+	params := &GetResponsesParams{PageSize: c.PageSize, After: token}
+	client := &http.Client{}
+
+	// GetTOKEN()
+	go func() {
+		defer close(events)
+		// TODO: for loop to paginate here?
+		params.After = token
+		for {
+			res, err := Call(client, c.BaseUrl, c.Key, form, params)
+
+			if err != nil {
+				handle(err)
+			}
+			for _, item := range res.Items {
+
+				// for pagination
+				params.After = item.Token
+
+				md := item.Hidden
+				// for _, dat := range item.Response {
+				// 	var ans Response
+				// 	err := json.Unmarshal(dat, &ans)
+				// 	if err != nil {
+				// 		handle(err)
+				// 	}
+
+				// rawAns, err := sjson.Delete(string(dat), "field")
+				if err != nil {
+					handle(err)
+				}
+
+				idx++
+				event := &InferenceDataEvent{
+					User:       User{ID: item.Token, Metadata: md},
+					Study:      source.StudyID,
+					SourceConf: source.Conf,
+					Timestamp:  item.Timestamp,
+					Idx:        idx,
+					Pagination: item.Token,
+					// Here I am not sure if it is the equivalent
+					Variable: item.Question_ref,
+					Value:    json.RawMessage(item.Question_text),
+				}
+				events <- event
+			}
+			// }
+
+			if res.TotalItems < params.PageSize {
```

Comment:
This looks great, but I don't see that the api is returning total_items! You'll either have to count the total items yourself or ask Sophie to include this.